src/backend/schemas/context.py

```diff
@@ -0,0 +1,154 @@
+from typing import Optional
+
+from fastapi import Request
+from pydantic import BaseModel, Field
+
+from backend.crud import user as user_crud
+from backend.database_models.database import DBSessionDep
+from backend.schemas.agent import Agent, AgentToolMetadata
+from backend.schemas.metrics import MetricsAgent, MetricsMessageType, MetricsUser
+from backend.schemas.user import User
+from backend.services.utils import get_deployment_config
+
+
+class Context(BaseModel):
+    request: Optional[dict] = {}
+    response: Optional[dict] = {}
+    receive: Optional[dict] = {}
+    trace_id: str = "default"
+    user_id: str = "default"
+    event_type: MetricsMessageType = None
+    user: Optional[User] = None
+    agent: Optional[Agent] = None
+    agent_tool_metadata: Optional[AgentToolMetadata] = None
+    model: Optional[str] = None
+    deployment_name: Optional[str] = None
+    deployment_config: Optional[dict] = None
+    conversation_id: Optional[str] = None
+    agent_id: Optional[str] = None
+
+    # Metrics
+    metrics_user: Optional[MetricsUser] = None
+    metrics_agent: Optional[MetricsAgent] = None
+
+    def set_request(self, request):
+        self.request = request
+
+    def set_response(self, response):
+        self.response = response
+
+    def set_receive(self, receive):
+        self.receive = receive
+
+    def with_trace_id(self, trace_id: str):
+        self.trace_id = trace_id
+
+    def with_user_id(self, user_id: str):
+        self.user_id = user_id
+
+    def with_deployment_name(self, deployment_name: str):
+        self.deployment_name = deployment_name
+
+    def with_event_type(self, event_type: MetricsMessageType) -> "Context":
+        self.event_type = event_type
+        return self
+
+    def with_user(
+        self, session: DBSessionDep | None = None, user: User | None = None
+    ) -> "Context":
+        if not user and not session:
+            return self
+
+        if not user:
+            user = user_crud.get_user(session, self.user_id)
+
+        if user:
+            self.metrics_user = MetricsUser(
+                id=user.id, email=user.email, fullname=user.fullname
+            )
+            self.user = user
+
+        return self
+
+    def with_agent(self, agent: Agent | None) -> "Context":
```

Comment:
should we use the same function to manage agent and metrics_agent? there also seem to be a short circuit when agent is not defined? should the function exit before defining `self.metrics_agent`?

Reply:
Sure, I can split this into two functions np

Reply:
Before my changes, we checked whether the agent existed, and if so, we added the metrics agent to the request state. Now, we don't check that before trying to add it to the context, so there's a chance we'll call `with_agent` with `agent = None`. In that case, we don't have any information to extract from the agent, and we can't define metrics_agent.

Is there a case in which we want to have only `agent` without having `metrics_agent`?

Reply:
I'd suggest separating the logic, re: https://github.com/cohere-ai/cohere-toolkit/pull/500#discussion_r1691885090

One approach would be:
-  create a new function: `with_metrics_agent(a: MetricsAgent)`
- When we want to track the metrics agent, we use the above function
- If agent doesn't exist, it's an error case, handle as needed
- If default agent is used (through new default agent endpoint), we call `with_metrics_agent(DEFAULT_METRICS_AGENT)`

Reply:
Oooooh okay, I get it now. Lemme make the changes